### ISSUE TITLE
Action plan #3

### ISSUE THREAD

**Issue by varshith15** [xlab-uiuc/slooo#9]
Please use the following code as template:
template(mongo): https://github.com/xlab-uiuc/slooo/blob/main/mongo/temp.xsh
server_configs: https://github.com/xlab-uiuc/slooo/blob/main/mongo/server_configs.json
super class to inherit from: https://github.com/xlab-uiuc/slooo/blob/main/utils/rsm.xsh

There are a lot of changes in https://github.com/xlab-uiuc/slooo/blob/main/utils/general.xsh
Please update your code carefully

Andrei:
- [ ] Update the redis code to support local mode
- [ ] Test redis code in both modes

Varshith:
- [x] Update the mongodb code to support local mode
- [x] Test mongodb code in both modes
- [x] Update the rethinkdb code to support local mode
- [x] Test rethinkdb code in both modes
- [x] Update the tidb code to support local mode
- [x] Test tidb code in both modes

Yuxuan:
- [x] Update the polardb code to support local mode
- [x] Test polardb code in both modes

**Comment by varshith15** [comment]
Summary for 27/09/21 meeting:

1. I updated the code to support local mode. Explained it to @TheEnthralled @Essoz 

- An example server_config file for reference ([mongo_server_config_file](https://github.com/xlab-uiuc/slooo/blob/main/mongo/server_configs.json)) to understand the update better. 

- The idea is to use a server_config json file to basically take all the options from, like IP, data directory, partition, port number, cpu to run on etc,. 

- For our code to work in a local mode some of those options are altered to suit the local setup like changing the IP to localhost, changing the data directory, port numbers, cpu number etc., so that there is no intersection between the servers as they are run on the same node in local mode.

- Previously these options(ports, data directories, cpu etc,) were hard coded as each server was being run on different node. 

- This gives the users a lot of options and they basically can run our code on their setup without changing any of our code they just need to update the the configs file

- Generating this server_config json is a one time work(For the first time when the experiment is run). So I don't think it should be inconvenient at all.

- The options in these server_configs are mostly common among all the DBs we have ported. But there are a few exceptions like mongo has specific option for logging_path which is not required in rethink or tidb. 

Prof. @tianyin what do you think of the above update.

2. I [implemented](https://github.com/xlab-uiuc/slooo/blob/main/utils/rsm.xsh) @TheEnthralled 's super class Idea.
3. @Essoz Has come up with a solution for the pgrep issue (mapping pids to servers when running in local mode)
4. Need to finish up the implementation soon, At current pace the agreed Oct 8th deadline won't be achieved.

**Comment by tianyin** [comment]
@varshith15 Those are great updates and great work everyone!

Could you show me two configs, one for local and one for cloud (taking mongo as the example)?

> The options in these server_configs are mostly common among all the DBs we have ported. But there are a few exceptions like mongo has specific option for logging_path which is not required in rethink or tidb.

So, how do you deal with common configs versus application-specific configs?

**Comment by varshith15** [comment]
> Could you show me two configs, one for local and one for cloud (taking mongo as the example)?

[local mode sever configs](https://github.com/xlab-uiuc/slooo/blob/main/mongo/local_server_configs.json) vs [remote mode server configs](https://github.com/xlab-uiuc/slooo/blob/main/mongo/server_configs.json) 

The difference is that the IPs, Data Directories, dbpaths, logpaths, cpu numbers, port numbers change accordingly to accomodate running of all the processes on the same node.

> So, how do you deal with common configs versus application-specific configs?

Currently my idea is to have all the configs(application-specific and common) in the same json file. So each DB will have a specific configs file. And for each DB we ported we provide a template json file that the users have to fill up manually before running their first experiment.

I'm open to your suggestions on this Prof. @tianyin . I'm out of ideas as to how to make it more joint with other ported DBs. @Essoz @TheEnthralled Please add if you have any suggestions/ideas on this as well.

**Comment by varshith15** [comment]
Summary for 4/10/21 meeting:

1. Ported and Tested the code for Mongo and Rethink in both modes. Will Port and Test TiDB. 
2. @Essoz has raised a PR on PolarDB (I will review it and suggest changes).
3. @TheEnthralled Is close to getting done with redis, he will raise a PR soon.

Have to push to get version 1 over the line this week.

**Comment by varshith15** [comment]
Summary for 14/10/21 meeting:

1. Porting/testing Mongo, Rethink, TiDB, Polar is done!!!
2. We have a working version 1!
3. Now its full focus on the paper!
Prof. @tianyin ^^

**Comment by tianyin** [comment]
@varshith15 

> We have a working version 1!

What do you mean by "working version"? :)

**Comment by varshith15** [comment]
Prof. @tianyin maybe working adjective wasn't necessary ;)
We have the version 1 of the tool ready. It's full tested on both local and cloud modes.

**Comment by tianyin** [comment]
> Prof. @tianyin 

Remove that stupid title please... it just sounds so awkward.

> We have the version 1 of the tool ready. It's full tested on both local and cloud modes.

I'm asking from the perspective of how you write your evaluation section.

**Comment by varshith15** [comment]
> Remove that stupid title please... it just sounds so awkward.

ok :)
 
> I'm asking from the perspective of how you write your evaluation section.

I don't think I understand your question clearly.

**Comment by tianyin** [comment]
You will need to write a section called "Evaluation" which basically tells what your tool can do and what results it can produce. What are those?

**Comment by varshith15** [comment]
The tool can run experiments (fail-slow fault tolerance checks) on systems we have currently ported to show the differences in the latency and throughput values(no slow vs slow). These experiments can be run on any setup. 

This tool makes it easier for developers to run those experiments on any system because porting different system to this framework takes very less time as compared to writing all the scripts from the scratch, The tool in a way provides a blue print for that. 

We will take mongo as a case study to explain in detail about the tool. We also have the slowness inject results for mongo from my previous trials.

**Comment by tianyin** [comment]
> This tool makes it easier for developers to run those experiments on any system because porting different system to this framework takes very less time as compared to writing all the scripts from the scratch, The tool in a way provides a blue print for that.

We can and should measure this by LOC for each tool.

**Comment by tianyin** [comment]
> We will take mongo as a case study to explain in detail about the tool. We also have the slowness inject results for mongo from my previous trials.

Awesome

**Comment by varshith15** [comment]
> We can and should measure this by LOC for each tool.

LOC?

**Comment by tianyin** [comment]
lines of code

**Comment by varshith15** [comment]
@tianyin By when do you think the initial draft of the paper should be ready by? Just to figure out a plan to deliver it by then.

**Comment by tianyin** [comment]
Which deadline are you targeting?

**Comment by Essoz** [comment]
> Which deadline are you targeting?

@tianyin 
the Nov 8 one

**Comment by tianyin** [comment]
Sorry for being able to respond earlier. We were running a conference last week.

Nov. 8? That's 8 days away. Where is your draft?

**Comment by Essoz** [comment]
@tianyin Had to say that we don't have a draft yet. But we have a rough outline of the paper [here](https://docs.google.com/document/d/1u_nEaePB8m4SOTCIqg22EWqvFMWzS44dliqItxX-okA/edit). Varshith and I are just about to start working on the draft.

**Comment by Essoz** [comment]
@tianyin 

Update: 11/4/2021

Varshith and I had a sync-up meeting today. We have some updates on our progress and some questions:

Varshith is about to finish the copilot part of the framework by tomorrow. The paper will have 4 parts : (Introduction, Interface, Design & Implementation, Framework Evaluation). Varshith picked the Interface part while I picked the Design & Implementation part. Hopefully we get a complete draft by 11/7/2021.

Honestly speaking, we are confused about the specific deadline of the paper and which conference are we targeting. Since it's the first time doing research for both of us, can you elaborate on all the deadline and conference stuff. We appreciate your help!

**Comment by tianyin** [comment]
@Essoz @varshith15 

Thanks for the diligence! 

> The paper will have 4 parts : (Introduction, Interface, Design & Implementation, Framework Evaluation

You will need the Related Work and Conclusion at least :P

Academic papers have a very specific format and I can help comment your drafts once you have a draft.

> Varshith picked the Interface part while I picked the Design & Implementation part

Yes, as I stressed in #6 , you should focus on these two sections.

> Hopefully we get a complete draft by 11/7/2021.

TBH, this does not sound real, but I'm looking forward. Paper writing takes a lot of time. 

> we are confused about the specific deadline of the paper and which conference are we targeting

The original plan is to submit to a tool/demo session of a software engineering conference. It seems that the closest one is this,
https://conf.researchr.org/track/icse-2022/icse-2022-demo---demonstrations

The deadline is Nov 24, 16 days away.

>  it's the first time doing research for both of us

Yes, it's a challenge! Honestly, writing a paper, even a 4-page one, takes a lot of time!! So my sense is that the ICSE deadline is hard. But there are many tool/demo submission opportunities along the way. My honest advice is to not be stressed about the deadlines; instead, START TO WRITE! Once you have a draft, we can discuss how good (and bad) it is and can estimate how much time we will need to deliver a strong submission. Now, I simply can't tell whether you need 4 days or 40 days.

**Comment by varshith15** [comment]
@tianyin 

Update: 24/11/21

@Essoz and I have been very busy with regular job / class work So we couldn't really work on the paper. It seems that way for the next week as well for @Essoz after which we plan to put more effort into finishing the paper up.

Meanwhile I was able to port copilot to our framework. The copilot code had many options (like durable, check, reply) which are not really explained in any of the documentation they provided. So I just left the options I didn't understand as default.
I have tested the ported code but I still haven't run any slow inject experiments yet to test if the copilot idea actually works.
